active_discussion/embedded/critical-section.md

```diff
@@ -0,0 +1,290 @@
+implements the `Send` trait.
+
+Example that shows that the bound is required: `type Type = Rc<u8>` could
+result in an unsound program (data race between `main` and `INTERRUPT0`).
+
+"Why are the memory clobbers required?" Without them the compiler can reorder
+`main`'s operations on `X` to outside the critical section leading to a data
+race.
+
+## Interrupt masking
+
+Consider this program that creates a critical section by masking a single
+interrupt (individual mask).
+
+> Aside: it's also possible to implement a critical section by raising the
+> running priority but the implementation of that kind of critical section is
+> very similar to this one (volatile write + compiler fence) so we won't include
+> it in this document.
+
+``` rust
+#![no_std]
+
+use core::{cell::UnsafeCell, ptr, sync::atomic::{self, Ordering}};
+
+extern "C" {
+    static MASK_INTERRUPT: UnsafeCell<u8>;
+    static UNMASK_INTERRUPT: UnsafeCell<u8>;
+}
+
+static mut X: Type = Type::default();
+
+#[no_mangle]
+unsafe fn main() -> ! {
+    unsafe {
+        asm!("ENABLE_INTERRUPTS" : : : : "volatile");
+    }
+
+    loop {
+        // .. any safe code ..
+
+        unsafe {
+            // start of critical section
+            ptr::write_volatile(MASK_INTERRUPT.get(), 1 << 0);
+        }
+
+        atomic::compiler_fence(Ordering::SeqCst);
+
+        // `INTERRUPT0` can *not* preempt this block
+        // (because it's masked)
+        {
+            let x: &mut Type = unsafe {
+                &mut X
+            };
+
+            // .. any safe code ..
+        }
+
+        atomic::compiler_fence(Ordering::SeqCst);
+
+        unsafe {
+            // end of critical section
+            ptr::write_volatile(UNMASK_INTERRUPT.get(), 1 << 0);
+        }
+
+        // .. any safe code ..
+    }
+}
+
+#[no_mangle]
+unsafe fn INTERRUPT0() {
+    let x: &mut Type = unsafe {
+        &mut X
+    };
+
+    // .. any safe code ..
+}
+```
+
+**Claim**:  This program is well-defined / sound if and only if `Type`
+implements the `Send` trait.
+
+Example that shows that the bound is required: `type Type = Rc<u8>` could
+result in an unsound program (data race between `main` and `INTERRUPT0`).
+
+"Why are the compiler fences required?" Without them the compiler can reorder
+`main`'s operations on `X` to outside the critical section leading to a data
+race.
+
+## Questions
+
+- Can these programs be misoptimized by the compiler? In particular, the
+  compiler fences in the second program prevent memory operations on `X` from
+  being reordered to outside the critical section but AFAIK they don't tell the
+  compiler that `X` may change outside the critical section -- could the program
+  cache the value of `X` on the stack? That would change the semantics of the
+  program.
+
+- I have observed that an `asm!("")` expression with *no* clobbers prevents
+  operations on `static mut` variables from being merged and reordered. See
+  example below. Is this intended behavior?
```

Comment:
I don't think that this is intended. LLVM may apply stronger constraints than necessary on asm, but that is not something that we should rely on.

Reply:
> that is not something that we should rely on

Rather than rely on this I would actually like to see instructions like `asm!("NOP" :::: "volatile")` *not* preventing optimizations.

(At some point I thought that the `sideeffect` attribute that's attached to to the `call asm` in the LLVM-IR was preventing the optimizations but after double checking today by running `opt` on the IR after removing the `sideeffect` attribute I see that that doesn't help.)

Reply:
That's probably related to [this issue](https://github.com/rust-lang/rfcs/pull/2360#issuecomment-466607141): LLVM is weird and expects the `memory` clobber to be partly handled by the frontend, by attaching `readnone`/`readonly` to the `call asm` instructions; Clang does this but rustc doesn't, so in effect, every Rust `asm!` statement is treated as if it had the `memory` clobber.

But also note that [Clang (but not GCC)](https://gcc.godbolt.org/z/u9AIvE) treats `volatile` as implying a `memory` clobber, and both Clang and GCC treat any `asm` statement with no output operands as [implicitly `volatile`](https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html#Volatile).  Whether Rust `asm!` does the same is... not fully defined, I guess; rustc does mark `asm!` without output operands as `sideeffect`, but the current implementation doesn't have an opinion on whether volatile should imply `memory`, since it treats everything as `memory`. :p